cloudinstall/charms/__init__.py

```diff
@@ -364,15 +356,33 @@ def watch_deploy(self):
                 self.ui.status_error_message(msg)
             time.sleep(10)
 
+    def filter_valid_relations(self):
+        deployed_charms_names = [x.charm_name for x in self.deployed_charms]
+        valid_relations = []
+        for c in self.deployed_charms:
+            for rel_a, rel_b in c.related:
+                rel_a_name = rel_a.split(":")[0]
+                rel_b_name = rel_b.split(":")[0]
+                if rel_a_name in deployed_charms_names and \
+                   rel_b_name in deployed_charms_names:
+                    valid_relations.append((rel_a, rel_b))
+        return valid_relations
+
     @utils.async
     def watch_relations_async(self):
         self.watch_relations()
 
     def watch_relations(self):
+        """
+        Handle charm relations
+        """
+        for charm in self.deployed_charms:
+            self.charm_relations_q.put(charm)
         log.debug("Starting charm relations watcher.")
         while not self.charm_relations_q.empty():
             try:
                 charm = self.charm_relations_q.get()
+                charm.related = self.filter_valid_relations()
```

Comment:
hmmm, so each charm is having its related list updated to be a list of all the valid relations in the system, which I think means that we're going to try to set relations many more times than necessary, and most of them will fail because the relation is already set (i forget it juju fails or just silently ignores setting a preexisting relation)